5.0 update status codes (#21)
neo4j/docs-status-codes: modules/ROOT/pages/index.adoc

|===

== Client errors

This is a complete list of all client error status codes Neo4j may return, and what they mean.

[options="header", cols="<1m,<1"]
|===

| Status Code
| Description

| Neo.ClientError.Cluster.NotALeader
| The request cannot be processed by this server. Write requests can only be processed by the leader.

| Neo.ClientError.Cluster.Routing
| Unable to route the request to the appropriate server

| Neo.ClientError.Database.DatabaseNotFound
| The request referred to a database that does not exist.

| Neo.ClientError.Database.ExistingAliasFound
| The request referred to a database with an alias.

| Neo.ClientError.Database.ExistingDatabaseFound
| The request referred to a database that already exists.

| Neo.ClientError.Fabric.AccessMode
| The request could not be completed due to access mode violation

| Neo.ClientError.General.ForbiddenOnReadOnlyDatabase
| This is a read only database, writing or modifying the database is not allowed.

| Neo.ClientError.General.InvalidArguments
| The request contained fields that were empty or are not allowed.

| Neo.ClientError.General.TransactionOutOfMemoryError
|
The transaction used more memory than was allowed.

> The maximum allowed size for a transaction can be configured with 'dbms.memory.transaction.max_size' in the neo4j configuration (normally in 'conf/neo4j.conf' or, if you are using Neo4j Desktop, found through the user interface).

Reviewer: ```suggestion
The maximum allowed size for a transaction can be configured with `dbms.memory.transaction.max_size` in the Neo4j configuration file (normally in _conf/neo4j.conf_ or, if you are using Neo4j Desktop, found through the user interface).
```

| The statement refers to a non-existent entity.

| Neo.ClientError.Statement.ExternalResourceFailed
| Access to an external resource failed.

| Neo.ClientError.Statement.NotSystemDatabaseError
| This is an administration command and it should be executed against the system database.

| Neo.ClientError.Statement.ParameterMissing
| The statement refers to a parameter that was not provided in the request.

| Neo.ClientError.Statement.PropertyNotFound
| The statement refers to a non-existent property.

| Neo.ClientError.Statement.RuntimeUnsupportedError
| This query is not supported by the chosen runtime.

| Neo.ClientError.Statement.SemanticError
| The statement is syntactically valid, but expresses something that the database cannot do.

| Neo.ClientError.Statement.SyntaxError
| The statement contains invalid or unsupported syntax.

| Neo.ClientError.Statement.TypeError
| The statement is attempting to perform operations on values with types that are not supported by the operation.

| Neo.ClientError.Statement.UnsupportedAdministrationCommand
| This administration command is not supported.

| Neo.ClientError.Statement.UnsupportedOperationError

> | This query preformed an operation that is not supported in this context.

Reviewer: ```suggestion
| This query performed an operation that is not supported in this context.
```

| Neo.ClientError.Transaction.ForbiddenDueToTransactionType
|
The transaction is of the wrong type to service the request.
For instance, a transaction that has had schema modifications performed in it cannot be used to subsequently perform data operations, and vice versa.

| Neo.ClientError.Transaction.InvalidBookmark
|
Supplied bookmark cannot be interpreted.
You should only supply a bookmark that was previously generated by Neo4j.
Maybe you have generated your own bookmark, or modified a bookmark since it was generated by Neo4j.

| Neo.ClientError.Transaction.InvalidBookmarkMixture
|
Mixing bookmarks generated by different databases is forbidden.
You should only chain bookmarks that are generated from the same database.
You may however chain bookmarks generated from system database with bookmarks from another database.

| Neo.ClientError.Transaction.LockClientStopped
| The transaction has been terminated, so no more locks can be acquired.

> This can occur because the transaction ran longer than the configured transaction timeout,
> or because a human operator manually terminated the transaction, or because the database is shutting down.

Reviewer: ```suggestion
This can occur because the transaction ran longer than the configured transaction timeout, or because a human operator manually terminated the transaction, or because the database is shutting down.
```

|===

== Notifications

This is a complete list of all notification status codes Neo4j may return, and what they mean.

[options="header", cols="<1m,<1"]
|===

| Status Code
| Description

| Neo.ClientNotification.Database.HomeDatabaseNotFound
| The request referred to a home database that does not exist.

| Neo.ClientNotification.Procedure.ProcedureWarning
| The query used a procedure that generated a warning.

| Neo.ClientNotification.Request.DeprecatedFormat
| The client made a request for a format which has been deprecated.

| Neo.ClientNotification.Schema.HintedIndexNotFound
| The request (directly or indirectly) referred to an index that does not exist.

| Neo.ClientNotification.Statement.CartesianProduct
| This query builds a cartesian product between disconnected patterns.

| Neo.ClientNotification.Statement.CodeGenerationFailed
| The database was unable to generate code for the query. A stacktrace can be found in the debug.log.

| Neo.ClientNotification.Statement.DynamicProperty
| Queries using dynamic properties will use neither index seeks nor index scans for those properties.

| Neo.ClientNotification.Statement.EagerOperator

> | The execution plan for this query contains the Eager operator, which forces all dependent data to be materialized in main memory before proceeding

Reviewer: ```suggestion
| The execution plan for this query contains the Eager operator, which forces all dependent data to be materialized in main memory before proceeding.
```

| Neo.ClientNotification.Statement.ExhaustiveShortestPath
|
Exhaustive shortest path has been planned for your query that means that shortest path graph algorithm might not be used to find the shortest path.

> Hence an exhaustive enumeration of all paths might be used in order to find the requested shortest path.

Reviewer: ```suggestion
Hence, an exhaustive enumeration of all paths might be used in order to find the requested shortest path.
```

| The provided pattern is unbounded, consider adding an upper limit to the number of node hops.

| Neo.ClientNotification.Statement.UnknownLabelWarning
| The provided label is not in the database.

| Neo.ClientNotification.Statement.UnknownPropertyKeyWarning
| The provided property key is not in the database.

| Neo.ClientNotification.Statement.UnknownRelationshipTypeWarning
| The provided relationship type is not in the database.

|===

== Transient errors

This is a complete list of all transient error status codes Neo4j may return, and what they mean.

[options="header", cols="<1m,<1"]
|===

| Status Code
| Description

| Neo.TransientError.Cluster.ReplicationFailure
| Replication failure.

| Neo.TransientError.Database.DatabaseUnavailable
|
The database is not currently available to serve your request, refer to the database logs for more details.
Retrying your request at a later time may succeed.

| Neo.TransientError.General.MemoryPoolOutOfMemoryError
|
The memory pool limit was exceeded.
The corresponding setting can be found in the error message.

| Neo.TransientError.General.OutOfMemoryError
|
There is not enough memory to perform the current task.

> Please try increasing `dbms.memory.heap.max_size` in the neo4j configuration (normally in `conf/neo4j.conf` or, if you are using Neo4j Desktop, found through the user interface) or if you are running an embedded installation increase the heap by using `-Xmx` command line flag, and then restart the database.

Reviewer: ```suggestion
Please try increasing `dbms.memory.heap.max_size` in the Neo4j configuration file (normally in _conf/neo4j.conf_ or, if you are using Neo4j Desktop, found through the user interface).

If you are running an embedded installation, you can increase the heap by using the `-Xmx` command line flag, and then restart the database.
```

| Neo.TransientError.General.StackOverFlowError
|
There is not enough stack size to perform the current task.

> This is generally considered to be a database error, so please contact Neo4j support.
> You could try increasing the stack size: for example to set the stack size to `2M`, add `dbms.jvm.additional=-Xss2M` to in the neo4j configuration (normally in `conf/neo4j.conf` or, if you are using Neo4j Desktop, found through the user interface) or if you are running an embedded installation just add `-Xss2M` as command line flag.

Reviewer: ```suggestion
This is generally considered to be a database error, so please contact Neo4j support.

You could try increasing the stack size; for example to set the stack size to `2M`, add `dbms.jvm.additional=-Xss2M` to in the Neo4j configuration file (normally in _conf/neo4j.conf_ or, if you are using Neo4j Desktop, found through the user interface).

 If you are running an embedded installation, you can add `-Xss2M` as command line flag.
```

| Neo.TransientError.General.TransactionMemoryLimit
|
There is not enough memory to perform the current task.

> Please try increasing `dbms.memory.off_heap.max_size` in the neo4j configuration (normally in `conf/neo4j.conf` or, if you are using Neo4j Desktop, found through the user interface), and then restart the database.

Reviewer: ```suggestion
Please try increasing `dbms.memory.off_heap.max_size` in the Neo4j configuration file (normally in _conf/neo4j.conf_ or, if you are using Neo4j Desktop, found through the user interface), and then restart the database.
```

| Neo.TransientError.Request.NoThreadsAvailable
|
There are no available threads to serve this request at the moment.

> You can retry at a later time or consider increasing max thread pool size for bolt connector(s).

Reviewer: ```suggestion
You can retry at a later time, or consider increasing max thread pool size for bolt connector(s).
```

| Neo.TransientError.Transaction.BookmarkTimeout
|
Bookmark wait timed out.
Database has not reached the specified version.

| Neo.TransientError.Transaction.ConstraintsChanged
| Database constraints changed since the start of this transaction.

| Neo.TransientError.Transaction.DeadlockDetected
|
This transaction, and at least one more transaction, has acquired locks in a way that it will wait indefinitely, and the database has aborted it.
Retrying this transaction will most likely be successful.

| Neo.TransientError.Transaction.Interrupted
| Interrupted while waiting.

| Neo.TransientError.Transaction.LeaseExpired
| The lease under which this transaction was started is no longer valid.

| Neo.TransientError.Transaction.LockAcquisitionTimeout
| Unable to acquire lock within configured timeout (`dbms.lock.acquisition.timeout`).

| Neo.TransientError.Transaction.MaximumTransactionLimitReached
|
Unable to start new transaction since the maximum number of concurrently executing transactions is reached (`dbms.transaction.concurrent.maximum`).
You can retry at a later time or consider increasing allowed maximum of concurrent transactions.

| Neo.TransientError.Transaction.Outdated
|

> Transaction has seen state which has been invalidated by applied updates while transaction was active.

Reviewer: ```suggestion
Transaction has seen state which has been invalidated by applied updates while the transaction was active.
```
